File: Cpp/thangmay.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int solve(int n, int k, int m, int s, int f) {
    vector<bool> visited(n + 1, false);
    queue<pair<int, int>> q;
    q.push({s, 0});
    while (!q.empty()) {
        pair<int, int> current = q.front();
        q.pop();
        if (current.first == f) return current.second;
        vector<int> next_floors = {current.first - m, current.first + k};
        for (int next_floor: next_floors) {
            if (1 <= next_floor && next_floor <= n && !visited[next_floor]) {
                visited[next_floor] = true;
                q.push({next_floor, current.second + 1});
            }
        }
    }
    return -1;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int n, k, m, s, f;
    cin >> n >> k >> m >> s >> f;
    cout << solve(n, k, m, s, f);
    return 0;
}

// code by duck it1

File: Python/thangmay.py
from collections import deque

def solve(n, k, m, s, f):
    visited = [False] * (n+1)
    queue = deque([(s, 0)])

    while queue:
        current_floor, steps = queue.popleft()
        if current_floor == f:
            return steps

        next_floors = [current_floor - m, current_floor + k]
        for next_floor in next_floors:
            if 1 <= next_floor <= n and not visited[next_floor]:
                visited[next_floor] = True
                queue.append((next_floor, steps + 1))

    return -1

n = int(input())
k, m = map(int, input().split())
s, f = map(int, input().split())
print(solve(n, k, m, s, f))